base/ltshipout.dtx

```diff
@@ -542,6 +556,33 @@
 %
 %
 %
+%  \begin{macro}{\@@_finalize_box:}
+%    For Lua\TeX{} invoke the \texttt{pre\_shipout\_filter} callback.
+%    \begin{macrocode}
+\sys_if_engine_luatex:TF
+  {
+    \newluafunction \@@_finalize_box:
+    \exp_args:Nx \everyjob {
+      \exp_not:V \everyjob
+      \exp_not:N \lua_now:n {
+        luatexbase.create_callback('pre_shipout_filter', 'list')
+        local~call, getbox, setbox = luatexbase.call_callback, tex.getbox, tex.setbox~
+        lua.get_functions_table()[\the \@@_finalize_box:] = function()
+          local~result = call('pre_shipout_filter', getbox(\the \l_shipout_box))
+          if~not (result == true) then~
+            setbox(\the \l_shipout_box, result~or~nil)
+          end~
+        end
+      }
+    }
+    \protected \luadef \@@_finalize_box: \the \@@_finalize_box:
+  } {
+    \cs_set_eq:NN \@@_finalize_box: \scan_stop:
+  }
```

Comment:
As a one-off I thin this would be fine, but if we expect other such callback to come up we might want to think about how to encapsulate that kind of setup differently (just a thought)

Reply:
I agree.

At this point I didn't copy the pseudo-primitive infrastructure from expl3, but that would already handle the `\newluafunction`, `everyjob` construction, `get_functions_table` and `luadef` parts. (It would also probably reuires  this to go to some Lua file or making some interfaces public.

This leaves the `create_callback` (which can't really be simplified, especially since other uses might need default functions here) and the actual `call(...)` / result handling which are quite specific to the individual usecase.

So I guess we could just adopt the expl3 model if necessary.